qiskit/transpiler/passes/utils/gates_basis.py

```diff
@@ -13,27 +13,52 @@
 """Check if all gates in the DAGCircuit are in the specified basis gates."""
 
 from qiskit.transpiler.basepasses import AnalysisPass
+from qiskit.transpiler.exceptions import TranspilerError
 
 
 class GatesInBasis(AnalysisPass):
     """Check if all gates in a DAG are in a given set of gates"""
 
-    def __init__(self, basis_gates):
+    def __init__(self, basis_gates=None, target=None):
         """Initialize the GatesInBasis pass.
 
         Args:
             basis_gates (list): The list of strings representing the set of basis gates.
+            target (Target): The target representing the backend. If specified
+                this will be used instead of the ``basis_gates`` parameter
+
+        Raises:
+            TranspilerError: If neither basis_gates or target is set.
         """
         super().__init__()
-        self._basis_gates = set(basis_gates)
+        if basis_gates is None and target is None:
+            raise TranspilerError(
+                "A value for 'basis_gates' or 'target' must be set to use this pass"
+            )
+        if basis_gates is not None:
+            self._basis_gates = set(basis_gates).union(
+                {"measure", "reset", "barrier", "snapshot", "delay"}
+            )
+        self._target = target
 
     def run(self, dag):
         """Run the GatesInBasis pass on `dag`."""
         gates_out_of_basis = False
-        basic_instrs = {"measure", "reset", "barrier", "snapshot", "delay"}
-        for gate in dag._op_names:
-            if gate not in self._basis_gates and gate not in basic_instrs:
-                gates_out_of_basis = True
-                break
-
+        if self._target is not None:
+            qubit_map = {qubit: index for index, qubit in enumerate(dag.qubits)}
+            for gate in dag.op_nodes():
+                # Barrier is universal and supported by all backends
+                if gate.name == "barrier":
+                    continue
+                if gate.name not in self._target:
+                    gates_out_of_basis = True
+                    break
+                if tuple(qubit_map[bit] for bit in gate.qargs) not in self._target[gate.name]:
+                    gates_out_of_basis = True
+                    break
```

Comment:
This might be digging too much into the internals of `Target`. Maybe an interface to query `Target` would be useful. Something like `self._target.gate_supported("cx", (0, 1))` or something like that. I imagine eventually querying `Target` will be useful elsewhere. OTOH, that might take more effort than we want to spend now. You could go back to `GatesInBasis` later and change it to use such a facility.

Reply:
Yeah, I think we should leave that for a separate PR as that's adding a new interface to the target class. We can keep this one contained to just improving the pass to support working with a target. Can you enhancement issue for this so we can track that for 0.20.0 too?

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/7548/commits/bdaba566837904c2509a529f75bdb6accef8c2a5 (I did it in another PR).